### PR TITLE
Enhance and fix resumable and ajax queues

### LINKED ISSUE #1557: Ajax resumable upload endless loop, (and multple chunks file upload issue ?)
## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [x] The issue still exists against the latest `master` branch of bootstrap-fileinput.
- [x] This is not an usage question. I confirm having read the plugin [documentation](http://plugins.krajee.com/file-input) and [demos](http://plugins.krajee.com/file-input/demo).
- [x] This is not a general programming / coding question. (Those should be directed to the [webtips Q & A forum](http://webtips.krajee.com/questions)).
- [x] I have attempted to find the simplest possible steps to reproduce the issue.
- [ ] I have included a failing test as a pull request (Optional).

## Steps to reproduce the issue

~I have not found a sure way to reproduce this yet, but here is my config~
**This can happen on demo page just upload a large file to force chunk upload, so larger than 2MB**

    $('#uploadInput').fileinput({
        language: uploadLocale,
        uploadAsync: true,
        enableResumableUpload: true,

        uploadUrl: actions.upload.url,
        uploadExtraData : {
            _token: 'tokenString',
            filePath: "${decodeURIComponent(getUrlLocationParameter('path'))}/"
        },
            
        showCaption: false, 
        showBrowse: false, 
        showPreview: true,
        showRemove: false, 
        showUpload: true, 
        showUploadStats: true,
        showCancel: null,
        showPause: null,
        showClose: false,
        showUploadedThumbs: true,
        showConsoleLogs: true,

        browseOnZoneClick: false,
        autoReplace: false,
        autoOrientImage: false, 
        autoOrientImageInitial: true,
        focusCaptionOnBrowse: true,
        focusCaptionOnClear: true,
        required: false,

        rtl: false,
        hideThumbnailContent: false,
        encodeUrl: true,
        generateFileId: null,
        previewClass: '',
        captionClass: '',
        frameClass: 'krajee-default',
        mainClass: 'file-caption-main',

        theme: "krajee-explorer",

        overwriteInitial: false,
        initialPreviewAsData: true,
        maxFileSize: 10000,
        removeFromPreviewOnError: true,
        previewFileType: 'any',

        fileSizeGetter: function(bytes) {
            i = Math.floor(Math.log(bytes) / Math.log(1024))
            sizes = [trans('B'), trans('KB'), trans('MB'), trans('GB'), trans('TB'), trans('PB'), trans('EB'), trans('ZB'), trans('YB')]
            return (bytes / Math.pow(1024, i)).toFixed(2) * 1 + ' ' + sizes[i]
        }
    });

The server answer

Capture d’écran de 2020-05-12 13-29-40

1. Setup a resumable upoad
2.  open the JS debugger
3. set a breakpoint on the if statement of  the processQueu method

 
        processQueue = function () {
            var config, xhr;
            if (self.ajaxCurrentThreads < self.maxAjaxThreads) {
                config = self.ajaxQueue.shift();
                if (typeof config !== 'undefined') {
                    self.ajaxCurrentThreads++;
                    xhr = $.ajax(config).done(function () {
                        clearInterval(self.ajaxQueueIntervalId);
                        self.ajaxCurrentThreads--;
                    });
                    self.ajaxRequests.push(xhr);
                }
            }
        };
## Expected behavior and actual behavior

This is the context, when i set the breakpoint: file has been uploaded, see the capture

Capture d’écran de 2020-05-12 13-38-15

**As you can see,  server answered "{\"chunkIndex\":\"4\",\"append\":true}"**

I believe this is wanted as stated in the example https://plugins.krajee.com/file-input#ajax-resumable

I was expecting...

AllTimeouts to be ended.

Also if i anwer HTTP 500 and error message from the server see the following capture, it seems to be stuck. The loop still happens.

Capture d’écran de 2020-05-12 13-42-58

## Environment

Browsers

- [x] Chromium Version 81.0.4044.138 (Build officiel) snap (64 bits)
- [ ] Google Chrome
- [x] Mozilla Firefox 76.0 (64 bits)
- [ ] Internet Explorer
- [ ] Safari

Operating System

- [ ] Windows
- [ ] Mac OS X
- [x] Linux
- [ ] Mobile

Libraries

- jQuery version:  3.5.1
- bootstrap-fileinput version:  5.0.9

## Isolating the problem

- [x] This bug happens [on the plugin demos page](http://plugins.krajee.com/file-input/demo)
- [x] The bug happens consistently across all tested browsers
- [ ] This bug happens when using bootstrap-fileinput without other plugins
- [ ] I can reproduce this bug in [a jsbin](https://jsbin.com/)

### PR DESCRIPTION
**Pull request by Kwaadpepper**
## Scope
This pull request includes a

- [x] Bug fix
- [ ] New feature
- [ ] Translation

## Changes
The following changes were made

- changed fileManager.processed to fileManager.filesProcessed
  this makes more sense, for code readability
  this is an optional change

- changed Object.keys to $h.getObjectKeys method
  this is an optional change
  https://stackoverflow.com/questions/5223/length-of-a-javascript-object

- added getObjectKeys method in $h
- added getObjectSize method in $h
- added taskManager in $h
- added $.whenAll jquery extension to handle running all chunks a once if
  resumableUploadOptions.maxThreads is equal to 0
  I think this could be removed, since we could want
  chunks to always be uploaded "maxThreads" at a time.
  If so some code in taskManager.TasksPool.runAllTasksParallels
  could be remove the "// if run all at once" part with the jquery
  extension

- ajax requests are stand alone tasks (promises)
- resumable uploads are a pool of tasks foreach chunks
  they are runned in parrallel

I wont remove the last setInterval for upload: function
this is too much work, and code readability is poor in my opinion.
It would benefit to be reactored using an object approach
instead of a functionnal programming.

There should also be a unique upload method: resumable.
It would be adapted for each options

## Related Issues
#1557

### CODE REVIEW

**Comment by kartik-v** [comment]
Thanks - I'll just optimize this a bit and post updated code over this PR.

**Comment by Kwaadpepper** [comment]
I did not remove all interval use.
This should be the purpose if this PR.

**Comment by kartik-v** [comment]
Yes - are you working on that? Let know... I thought I could optimize the Task manager code a bit.

**Comment by Kwaadpepper** [comment]
Yes, my bad. I will let tout know when its ready.

**Comment by Kwaadpepper** [comment]
Think this is ready for you to check.
I think there would be some improvement to be done with the progress bar.
I left some comments.

**Comment by kartik-v** [comment]
Thanks. I am checking this and have tried to optimize the TaskManager lib. The initial issue I am seeing is that *pause* and *resume* functionality seems broken somewhere as well as updating of the main progress bar. Will need to check and test in detail if anything else.

**Comment by kartik-v** [comment]
There needs to be a pause and resume feature - need to check your task pool implementation - currently it has only cancel and success. Checking on this.

**Comment by Kwaadpepper** [comment]
You're right looks like there is an issue with resuming. It restarts uploading all chunks. As for the main progress bar i'm not sure what is causing this issue i tried to keep original code as much as possible. I need to check where it is updated.

Checking this.

**Comment by Kwaadpepper** [comment]
Ok i fixed things for resume uploads. I tested resume. It seems ok to me.

I noticed a bug though. When an upload starts for multiple files. I you pause, then browse and add a file. It's size is not added to ```fm.totalSize```, So the general progress bar percent calculation will be at 100% when this last added file will be starting to be uploaded. This is a new bug unrelated to this PR. I checked i can reproduce this on your example https://plugins.krajee.com/file-resumable-uploads-demo#resumable-uploads-1.

**Comment by Kwaadpepper** [comment]
As for tasks. It can't be paused of course as well as Ajax query since they are promises. But undone chunks can be re-added to the queue. This is the actual behavior. So there is no pause related code in task manager.

**Comment by kartik-v** [comment]
> I you want so I can squash the PR.

Thanks - you can.

**Comment by Kwaadpepper** [comment]
done.

**Comment by kartik-v** [comment]
> As for the main progress bar i'm not sure what is causing this issue i tried to keep original code as much as possible. I need to check where it is updated.

Thanks. I am not finding issues with the main progress bar updating in the new code ... can you let know - what is the issue you are facing with the main progress bar update?

Will check the other parts in the meanwhile - adding a new file to a paused resumable.

**Comment by kartik-v** [comment]
I think the task manager code in itself can be optimized to a more compressed one than it exists now. I will do it myself at the end once we get the others tested.

**Comment by Kwaadpepper** [comment]
> Thanks. I am not finding issues with the main progress bar updating in the new code ... can you let know - what is the issue you are facing with the main progress bar update?

My last commit fixed the main progress bar issue i had.


> I think the task manager code in itself can be optimized to a more compressed one than it exists now. I will do it myself at the end once we get the others tested.

Sure, there is code we can pull out. For example. the jquey plugin introducing $.whenall is not necessary if you don't allow max thread to be 0 ( upload all at once). I wrote code to handle that. But if you consider there is always a max number and a min number for ajax queries to run in parallel than you can remove $.whenAll plugin and if statement on line 660: runAllTasksParallels method.

File: js/fileinput.js [hunk 1]
```diff
@@ -588,6 +588,222 @@
...
+                        doneTasksListResults = [],
+                        deferred = $.Deferred();
+
+                    if (self.cancelTasks) {
+                        self.cancelTasksDeferrer.resolve();
+                        return deferred.reject();
+                    }
+
+                    // if run all at once
+                    if (!maxParallelsTaskNumber) {
```
Comment (Kwaadpepper) [inline js/fileinput.js:661]: This could be removed eventually

File: js/fileinput.js [hunk 2]
```diff
@@ -5722,6 +5941,58 @@
 
     $.fn.fileinput.Constructor = FileInput;
 
+    /**
+     * Small dependency injection for the task manager
+     * https://gist.github.com/fearphage/4341799
+     */
+    $.whenAll = function(array) {
```
Comment (Kwaadpepper) [inline js/fileinput.js:5948]: This could be removed eventually

File: js/fileinput.js [hunk 3]
```diff
@@ -1261,17 +1487,32 @@
...
+
+                        // use fifo chunk stack
+                        var arr = rm.stack.shift(), index;
+
+                        index = arr[0];
+                        if (!rm.chunksProcessed[rm.id] || !rm.chunksProcessed[rm.id][index]) {
+                            rm.sendAjax(index, arr[1], deferrer);
+                            return;
+                        } else {
+                            console.log('wtf');
```
Comment (kartik-v) [inline js/fileinput.js:1502]: There is one issue I am seeing on the pause and resume. When you pause it once and then resume it works fine. But when you pause and resume more than once for the same file - the code runs into an issue and enters this console.log... need to check.

**Comment by Kwaadpepper** [comment]
I am seeing some issues on my project that needs to be inverstigated:
Events are triggered when cancelling file upload (resumable) such as 

    bootstrap-fileinput: "uploadInput": 0: abort. Error Details: . fileinput.js:2153
    filechunkajaxerror
    filechunkcomplete

This is because hxr promise is rejected ( calling abort method ), It is not handled properly.

File: js/fileinput.js (hunk 3 above)
Comment (Kwaadpepper) [reply js/fileinput.js:1502]: OKay i'll look at that.

**Comment by kartik-v** [comment]
Yes abort of xhr is expected in case of cancel ... when one will not resume.., but it should not when one is pausing... the resumableManager init must not happen during pause and resume unlike how its during cancel or a fresh upload. 

So interestingly --- I took up your PR modification - you did on intervals ... the last updated one before you introduced the queue...  I modified that code a bit for pause and resume...seems to work well for most part... only need to check the new file added during pause state...

**Comment by Kwaadpepper** [comment]
Yes i could make it work using interval, but it seems to me it wasn't a good solution. I feel like using promises is a more neat solution. It is up to you after all to decide what you want to use.

> Yes abort of xhr is expected in case of cancel ... to terminate the ajax requests in process...,

This should not raise ```filechunkajaxerror``` or ```filechunkcomplete``` event in the process that was my point

**Comment by Kwaadpepper** [comment]
> There is one issue I am seeing on the pause and resume. When you pause it once and then resume it works fine. But when

```rm.stack``` is growing more and more when pause and resume few times this is the cause. 
I will fix this if you want to use the task manager solution. I you prefer to use intervals as my previous PR i will leave it as it is and let you finish this work. As you wish.

**Comment by Kwaadpepper** [comment]
>but for pause and resume I feel the resumableManager init must not happen unlike how it happens during cancel or a fresh upload.

I had this taught also.  It happens that you are calling reset only when a fresh upload would occur. this is not really intuitive. Imao, code could be restructured in a more object oriented style. It would be easier to maintain. Maybe using coffeescript ? But this is a lot of work.

**Comment by kartik-v** [comment]
Great. Will have a look at this.

**Comment by kartik-v** [comment]
> I had this taught also. It happens that you are calling reset only when a fresh upload would occur. this is not really intuitive. Imao, code could be restructured in a more object oriented style. It would be easier to maintain. Maybe using coffeescript ? But this is a lot of work.

I have been thinking of major revamp to the library as well . Initially this library was created as a bootstrap extension for file input and upload and features kept getting added with resumable uploads coming up recently. Meanwhile, we had bootstrap versions going from 2.x, then 3.x and now bootstrap 4.x and plugin changed to cater to these releases and throughout has been also dependent on the jQuery library. 

If you are interested we can collaborate on a new branch or even a new repo to create a rehash of this library with latest javascript or coffeescript so that it works across all other frameworks.

**Comment by Kwaadpepper** [comment]
Sure I would like to collab on my free time. I am also working on my own side project ( a laravel file manager ), so working on my free time.

>  a rehash of this library with latest javascript or coffeescript so that it works across all other frameworks.

Which specific framework are you referring to ?

I am a Php full stack dev, and I have some experience with C# and other compiled code, but I am not a professional JS dev. I would have some suggestions though. But please give us some guidelines first.

In the meantime would you like to finish this PR first ?

**Comment by kartik-v** [comment]
> Which specific framework are you referring to ?
Make it extensible to other javascript frameworks beyond jQuery (or make jQuery optional) if needed. Which means work with pure javascript - but it needs a bit of rethink. Anyway this is a separate project to take up and can be done in free time.

> In the meantime would you like to finish this PR first ?
Yes I have done the final changes on your code - and now it seems to work fine - but I am doing some testing on this at my end and will close the PR with the final code using your task manager logic if all goes fine.

**Comment by Kwaadpepper** [comment]
Nice.

>I meant, make it extensible to be used in other javascript frameworks beyond jQuery - e.g. react, angular etc.. (or make jQuery optional) if needed. Which means work with pure javascript - but it needs a bit of rethink.

I see, removing jQuery could be done in vanilla JS if course. But as you said before if we don't want to reinvent the wheel, the question would be which browser you want to support. I guess not under ie9 or may be ie10 ?

**Comment by kartik-v** [comment]
Enhancement done and pushed. I have taken and refactored your code and simplified / optimized  the queue functionality. Have a relook from your end on your testing and let know as well.

**Comment by Kwaadpepper** [comment]
I just pulled last master branch. I ran into this:
Capture d’écran de 2020-05-18 11-49-28

After an upload failed, the main progress bar was not updated. Server anwsered 500 for a chunk upload. I am looking at this.

**Comment by Kwaadpepper** [comment]
An other issue, when upload is finished, the main progress bar is not updating but stuck in 
Capture d’écran de 2020-05-18 12-05-00

**Comment by Kwaadpepper** [comment]
I think these issues are the same. All chunks must be synced when done, then only the UI is updated.

**Comment by Kwaadpepper** [comment]
Capture d’écran de 2020-05-18 12-25-12

**Comment by Kwaadpepper** [comment]
```sendAjax.{fnSuccess,fnError}``` methods should be aware that ```setProcessed``` was previously called, ot the other way around, ```setProcessed``` should not have been called till all chunks returned

**Comment by kartik-v** [comment]
Will check in sometime on this. please share your findings in the meanwhile.

**Comment by Kwaadpepper** [comment]
Solved the isse with this patch 

    diff --git a/js/fileinput.js b/js/fileinput.js
    index f185262..0976606 100755
    --- a/js/fileinput.js
    +++ b/js/fileinput.js
    @@ -696,7 +696,7 @@
                         tp.run = function (maxThreads) {
                             var i = 0, failed = false, task, tasksList = $h.getObjectKeys(tp.tasks).map(function (key) {
                                 return tp.tasks[key];
    -                        }), tasksCount = tasksList.length, tasksDone = [], deferred = $.Deferred(), enqueue, callback;
    +                        }), tasksDone = [], deferred = $.Deferred(), enqueue, callback;
     
                             if (tp.cancelled) {
                                 tp.cancelledDeferrer.resolve();
    @@ -751,7 +751,7 @@
                                     tp.cancelledDeferrer.resolve();
                                     return;
                                 }
    -                            if (tasksDone.length === tasksCount) {
    +                            if (tasksDone.length === tp.size()) {
                                     if (failed) {
                                         deferred.reject.apply(null, tasksDone);
                                     } else {

**Comment by Kwaadpepper** [comment]
I just ran through this also
Capture d’écran de 2020-05-18 13-41-16

When uploading, just cancel at any time to reproduce this.

**Comment by Kwaadpepper** [comment]
diff --git a/js/fileinput.js b/js/fileinput.js
    index f185262..f020296 100755
    --- a/js/fileinput.js
    +++ b/js/fileinput.js
    @@ -696,7 +696,7 @@
                         tp.run = function (maxThreads) {
                             var i = 0, failed = false, task, tasksList = $h.getObjectKeys(tp.tasks).map(function (key) {
                                 return tp.tasks[key];
    -                        }), tasksCount = tasksList.length, tasksDone = [], deferred = $.Deferred(), enqueue, callback;
    +                        }), tasksDone = [], deferred = $.Deferred(), enqueue, callback;
     
                             if (tp.cancelled) {
                                 tp.cancelledDeferrer.resolve();
    @@ -751,7 +751,7 @@
                                     tp.cancelledDeferrer.resolve();
                                     return;
                                 }
    -                            if (tasksDone.length === tasksCount) {
    +                            if (tasksDone.length === tp.size()) {
                                     if (failed) {
                                         deferred.reject.apply(null, tasksDone);
                                     } else {
    @@ -1504,6 +1504,10 @@
                             self._raise('filechunkbeforesend', [id, index, retry, fm, rm, outData]);
                         };
                         fnSuccess = function (data, textStatus, jqXHR) {
    +                        if (self.cancelling) {
    +                            deferrer.reject('cancelling');
    +                            return;
    +                        }
                             outData = self._getOutData(fd, jqXHR, data);
                             var paramNames = self.uploadParamNames, chunkIndex = paramNames.chunkIndex || 'chunkIndex',
                                 opts = self.resumableUploadOptions, params = [id, index, retry, fm, rm, outData];
    @@ -1531,6 +1535,10 @@
                             }
                         };
                         fnError = function (jqXHR, textStatus, errorThrown) {
    +                        if (self.cancelling) {
    +                            deferrer.reject('cancelling');
    +                            return;
    +                        }
                             outData = self._getOutData(fd, jqXHR);
                             rm.error = errorThrown;
                             rm.logAjaxError(jqXHR, textStatus, errorThrown);
    @@ -1541,7 +1549,9 @@
                             deferrer.reject('try failed');
                         };
                         fnComplete = function () {
    -                        self._raise('filechunkcomplete', [id, index, retry, fm, rm, self._getOutData(fd)]);
    +                        if (!self.cancelling) {
    +                            self._raise('filechunkcomplete', [id, index, retry, fm, rm, self._getOutData(fd)]);
    +                        }
                         };
                         self._ajaxSubmit(fnBefore, fnSuccess, fnComplete, fnError, fd, id, rm.fileIndex);
                     }

**Comment by Kwaadpepper** [comment]
This patch solves both to last issues, 
But the UI is not fully updated : 
Capture d’écran de 2020-05-18 14-07-01

**Comment by kartik-v** [comment]
Thanks. Can you share the changes over a new PR? Will fix the filechunk events for aborted scenarios.

**Comment by kartik-v** [comment]
Thanks. Can you share these changes over a new PR? 

I will fix the filechunk events for aborted scenarios via a separate issue tracking that.

**Comment by kartik-v** [comment]
All together fixed via #1561. Please check and report anything over that issue so we can track that. Thanks.

**Pull request closed without merging**